File: 백준/Gold/11559. Puyo Puyo/Puyo Puyo.cc
#include <bits/stdc++.h>
using namespace std;

typedef pair<int, int> pii;
const int dy[4] = {0, 1, 0, -1};
const int dx[4] = {1, 0, -1, 0};
const int H = 12, W = 6;

char arr[14][8];
int ans = 0;

bool bfs(int y, int x) {
  bool visited[14][8] = {};
  queue<pii> q;
  vector<pii> seq;

  q.push({y, x});
  seq.push_back({y, x});
  visited[y][x] = true;

  while (!q.empty()) {
    pii p = q.front();
    q.pop();
    for (int i = 0; i < 4; i++) {
      int ny = dy[i] + p.first;
      int nx = dx[i] + p.second;
      if (visited[ny][nx]) continue;
      if (arr[ny][nx] == '.') continue;
      if (arr[y][x] == arr[ny][nx]) {
        visited[ny][nx] = true;
        q.push({ny, nx});
        seq.push_back({ny, nx});
      }
    }
  }
  if (seq.size() < 4) return false;
  for (pii p : seq) {
    arr[p.first][p.second] = '.';
  }
  return true;
}

bool crack() {
  bool ret = false;
  for (int j = 1; j <= H; j++) {
    for (int i = 1; i <= W; i++) {
      if (arr[j][i] != '.' && bfs(j, i)) ret = true;
    }
  }
  return ret;
}

void falldown() {
  for (int i = 1; i <= W; i++) {
    for (int j = 1; j < H; j++) {
      if (arr[j][i] == '.') {
        for (int k = j + 1; k <= H; k++) {
          if (arr[k][i] != '.') {
            swap(arr[j][i], arr[k][i]);
            break;
          }
        }
      }
    }
  }
}

void solve() {
  while (crack()) {
    ans++;
    falldown();
  }
  cout << ans << "\n";
}

int main() {
  ios::sync_with_stdio(false), cin.tie(0), cout.tie(0); /* FastIO */

  fill_n(arr[0], 14 * 8, '.');
  for (int j = H; j >= 1; j--) {
    for (int i = 1; i <= W; i++) {
      cin >> arr[j][i];
    }
  }

  solve();
}

File: 백준/Gold/11559. Puyo Puyo/Puyo Puyo.py
from collections import deque

H, W = 12, 6
dy = [0, 1, 0, -1]
dx = [1, 0, -1, 0]

arr = [['.'] * (W + 2) for _ in range(H + 2)]
ans = 0


def bfs(y, x):
    visited = [[False] * (W + 2) for _ in range(H + 2)]
    q = deque()
    seq = []

    q.append((y, x))
    seq.append((y, x))
    visited[y][x] = True

    while q:
        cy, cx = q.popleft()
        for d in range(4):
            ny, nx = cy + dy[d], cx + dx[d]
            if visited[ny][nx]:
                continue
            if arr[ny][nx] == '.':
                continue
            if arr[ny][nx] == arr[y][x]:
                visited[ny][nx] = True
                q.append((ny, nx))
                seq.append((ny, nx))

    if len(seq) < 4:
        return False

    for py, px in seq:
        arr[py][px] = '.'
    return True


def crack():
    flag = False
    for y in range(1, H + 1):
        for x in range(1, W + 1):
            if arr[y][x] != '.' and bfs(y, x):
                flag = True
    return flag


def falldown():
    for x in range(1, W + 1):
        for y in range(1, H):
            if arr[y][x] == '.':
                for k in range(y + 1, H + 1):
                    if arr[k][x] != '.':
                        arr[y][x], arr[k][x] = arr[k][x], arr[y][x]
                        break


def solve():
    global ans
    while crack():
        ans += 1
        falldown()
    print(ans)


if __name__ == "__main__":
    for y in range(H, 0, -1):
        line = input()
        for x in range(1, W + 1):
            arr[y][x] = line[x - 1]

    solve()
